FWCore/Framework/interface/Callback.h

```diff
@@ -9,231 +9,80 @@
-                    runProducerAsync(group, iExcept, iRecord, iEventSetupImpl, weakToken.lock());
-                  }
-                });
-
-            //Get everything we can before knowing about the mayGets
-            prefetchNeededDataAsync(WaitingTaskHolder(*group, mayGetTask), iEventSetupImpl, getTokenIndices(), token);
-          } else {
-            ServiceWeakToken weakToken = token;
-            auto task = edm::make_waiting_task(
-                [this, group, iRecord, iEventSetupImpl, weakToken](std::exception_ptr const* iExcept) {
-                  runProducerAsync(group, iExcept, iRecord, iEventSetupImpl, weakToken.lock());
-                });
-            prefetchNeededDataAsync(WaitingTaskHolder(*group, task), iEventSetupImpl, getTokenIndices(), token);
-          }
-        }
-      }
-
-      template <class DataT>
-      void holdOntoPointer(DataT* iData) {
-        proxyData_[produce::find_index<TReturn, DataT>::value] = iData;
-      }
-
-      void storeReturnedValues(TReturn iReturn) {
-        using type = typename produce::product_traits<TReturn>::type;
-        setData<typename type::head_type, typename type::tail_type>(iReturn);
+        return Base::prefetchAsyncImpl(
+            [this](auto&& group, auto&& token, auto&& record, auto&& es) {
+              return makeRunModuleTaskHolder(group, token, record, es);
+            },
+            std::move(iTask),
+            iRecord,
+            iEventSetupImpl,
+            token,
+            iParent);
       }
 
-      template <class RemainingContainerT, class DataT, class ProductsT>
-      void setData(ProductsT& iProducts) {
-        DataT* temp = reinterpret_cast<DataT*>(proxyData_[produce::find_index<TReturn, DataT>::value]);
-        if (nullptr != temp) {
-          moveFromTo(iProducts, *temp);
-        }
-        if constexpr (not std::is_same_v<produce::Null, RemainingContainerT>) {
-          setData<typename RemainingContainerT::head_type, typename RemainingContainerT::tail_type>(iProducts);
-        }
-      }
-      void newRecordComing() {
-        wasCalledForThisRecord_ = false;
-        taskList_.reset();
-      }
-
-      unsigned int transitionID() const { return id_; }
-      ESProxyIndex const* getTokenIndices() const { return producer_->getTokenIndices(id_); }
-
     private:
-      void prefetchNeededDataAsync(WaitingTaskHolder task,
-                                   EventSetupImpl const* iImpl,
-                                   ESProxyIndex const* proxies,
-                                   edm::ServiceToken const& token) const {
-        auto recs = producer_->getTokenRecordIndices(id_);
-        auto n = producer_->numberOfTokenIndices(id_);
-        for (size_t i = 0; i != n; ++i) {
-          auto rec = iImpl->findImpl(recs[i]);
-          if (rec) {
-            rec->prefetchAsync(task, proxies[i], iImpl, token, edm::ESParentContext{&callingContext_});
-          }
-        }
+      Callback(T* iProd,
```

Comment:
One thing I just noticed, I made this constructor private and the other constructor now delegates to it because of comments from Matti (a general comment about making things private that could be private) and Chris (about delegating instead directly initializing the base class), but perhaps a better modification would be to delete this constructor. Nothing uses it. I only included it because it existed before. Is there some expectation that we will need this in the future or that code outside of the repository needs this? It's a very minor thing either way.

Also let me know when the review is complete and I'll squash all these commits.

Reply:
> but perhaps a better modification would be to delete this constructor. Nothing uses it.

Isn't this constructor being used by `clone()`? (or did I miss something?)

> Also let me know when the review is complete and I'll squash all these commits.

I want to take another look after your latest changes, I hope to get to it today.

Reply:
You're right. Sorry for the noise. It is used by clone.